### PR TITLE
ISPN-15651 HR Rolling Upgrades fail with ___protobuf_metadata cache

### PR DESCRIPTION
**Pull request by ryanemerson**
https://issues.redhat.com/browse/ISPN-15651

I still need to verify the fix with a built image in the Operator and cleanup the RollingUpgrade tests to reduce repetition and ensure the Protobuf cache is covered in all scenarios.

### CODE REVIEW

**Comment by infinispanrelease** [comment]
Image pushed for Jenkins build [#3](https://ci.infinispan.org/job/Infinispan/job/PR-11921/3/):
```
quay.io/infinispan-test/server:PR-11921
```

**Comment by infinispanrelease** [comment]
Image pushed for Jenkins build [#5](https://ci.infinispan.org/job/Infinispan/job/PR-11921/5/):
```
quay.io/infinispan-test/server:PR-11921
```

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java [hunk 1]
```diff
@@ -481,6 +481,8 @@ private boolean shouldIntercept(Object key) {
 
    @Override
    public Object visitComputeCommand(InvocationContext ctx, ComputeCommand command) {
+      if (command.hasAnyFlag(FlagBitSets.ROLLING_UPGRADE))
+         return invokeNext(ctx, command);
```
Comment (fax4ever) [inline remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:485]: I think that theoretically we should run a reindex or index update on all possibly impacted indexed caches that use the entities that have changed. 
Since I think we have never covered this case so far, I think that it is safe to log something.
We should consider to open a separate issue to address the topic here, that if don't get wrong, it is about to support reindex on rolling upgrades.

**Review by fax4ever — APPROVED** [review]
Nice job! I would say +1 to merge after introducing a log to alter the user that if there are indexed caches using the modified entities, they should probably reindexed.

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java (hunk 1 above)
Comment (ryanemerson) [reply remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:485]: Hmm that's an interesting point. In the case of a rolling upgrade in the Operator, the "___protobuf_metadata" cache is always sync'd from the source -> target cluster before the user caches in our testsuite as they all have lower-case names. If I change the cachename to upper case, then the TreeSet will order the cache before the "___protobuf_metadata" cache and cache creation will fail as entities won't yet exist on the target cluster. I've created https://github.com/infinispan/infinispan-operator/issues/1988 to address this.

If the Operator ensures that the "___protobuf_metadata" data is always sync'd before the user caches are created, I assume that there's no need or any of the user caches to be reindexed?

Comment (ryanemerson) [reply remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:485]: > Since I think we have never covered this case so far, I think that it is safe to log something.

My only issue with logging something is that in the Operator case, the user won't need to do anything due to the order of events described in my previous comment, so the warning logs will probably confuse the user as no action is actually required in that case.

Instead of logging a warning, how about we create a new issue to update the HR rolling upgrade docs so that we explicitly state that users with indexed caches must sync the "___protobuf_metadata" cache before other caches? Or in the case that they're migrating to an already partially populated cluster (I don't think this use-case is common) they have to ensure they re-index the cache after upgrade?

Comment (fax4ever) [reply remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java:485]: creating a new issue makes totally sense to me. thanks

**Comment by infinispanrelease** [comment]
Image pushed for Jenkins build [#6](https://ci.infinispan.org/job/Infinispan/job/PR-11921/6/):
```
quay.io/infinispan-test/server:PR-11921
```

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java (hunk 1 above)
Comment (ryanemerson) [reply remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java]: I've created [ISPN-15696](https://issues.redhat.com/browse/ISPN-15696) for the documentation and added a commit to this PR.

**Comment by ryanemerson** [comment]
@fax4ever I had to rework my original solution as the .proto files were never registered. The `ComputeCommand` is now implemented in the interceptor, but I have still limited this to the ROLLING_UPGRADE use-case as there might be some edge cases that I haven't addressed.

**Comment by infinispanrelease** [comment]
Image pushed for Jenkins build [#7](https://ci.infinispan.org/job/Infinispan/job/PR-11921/7/):
```
quay.io/infinispan-test/server:PR-11921
```

File: remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java (hunk 1 above)
Comment (fax4ever) [reply remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerInterceptor.java]: thanks!

**Comment by infinispanrelease** [comment]
Image pushed for Jenkins build [#8](https://ci.infinispan.org/job/Infinispan/job/PR-11921/8/):
```
quay.io/infinispan-test/server:PR-11921
```
